gaffer-core/graph/src/main/java/gaffer/graph/hook/OperationChainLimiter.java

```diff
@@ -0,0 +1,262 @@
+     */
+    public OperationChainLimiter(final Properties operationScorePropertiesFile, final Properties operationAuthorisationScoreLimitPropertiesFile) {
+        loadMapsFromProperties(operationScorePropertiesFile, operationAuthorisationScoreLimitPropertiesFile);
+    }
+
+    /**
+     * Checks the {@link OperationChain}
+     * is allowed to be executed by the user.
+     * This is done by checking the user's auths against the auth scores getting the users maximum score limit value.
+     * Then checking the operation score of all operations in the chain and comparing the total score value of the chain against a users maximum score limit.
+     * If an operation cannot be executed then an {@link IllegalAccessError} is thrown.
+     *
+     * @param user    the user to authorise.
+     * @param opChain the operation chain.
+     */
+    @Override
+    public void preExecute(final OperationChain<?> opChain, final User user) {
+        if (null != opChain) {
+            Integer chainScore = 0;
+            Integer maxRoleScore = getMaxUserRoleScore(user.getOpAuths());
+            for (Operation operation : opChain.getOperations()) {
+                chainScore += authorise(operation, user);
+                if (chainScore > maxRoleScore) {
+                    throw new UnauthorisedException("The maximum score limit for this user is " + maxRoleScore + ".\n" +
+                            "The requested operation chain exceeded this score limit.");
+                }
+            }
+        }
+    }
+
+    /**
+     * Iterates through each of the users operation authorisations listed in the config file and returns the highest score
+     * associated with those roles.
+     * <p/>
+     * Defaults to 0.
+     *
+     * @param opAuths a set of operation authorisations
+     * @return maxUserScore the highest score associated with any of the supplied user roles
+     */
+    private Integer getMaxUserRoleScore(final Set<String> opAuths) {
+        Integer maxUserScore = 0;
+        for (String opAuth : opAuths) {
+            Integer roleScore = authScores.get(opAuth);
+            if (null != roleScore) {
+                if (roleScore > maxUserScore) {
+                    maxUserScore = roleScore;
+                }
+            }
+        }
+        LOGGER.info("Returning users max operation chain limit score of " + maxUserScore);
+        return maxUserScore;
+    }
+
+    @Override
+    public void postExecute(final Object result, final OperationChain<?> opChain, final User user) {
+        // This method can be overridden to add additional authorisation checks on the results.
+    }
+
+    protected Integer authorise(final Operation operation, final User user) {
+        Integer operationScore = null;
+        if (null != operation) {
+            final Class<? extends Operation> opClass = operation.getClass();
+            for (Map.Entry<Class<? extends Operation>, Integer> entry : operationScores.entrySet()) {
```

Comment:
It would be more efficient to add a break when the first entry with assignable operation class is found.
If you need the last operation class then can you iterate through the operation scores in reverse order?